### DISCUSSION TITLE
To batch Tina commits

### DISCUSSION THREAD

**Discussion by clementroche** [tinacms/tinacms#4289] [Ideas]
Hey, by using TinaCMS on a project client we realised that it generates a lot of git commits, everytime you hit "Save" on production mode. In addition we're triggering a build on every deployment to make those changes live (using Vercel). But sometimes client edits hundreds of fields within 10 minutes which leads to gigantic list of commits and very long queues of builds, blocking almost eveything for hours.
I'm wondering if there is a way to "batch" commits done in one session and only push it at one time when nothing was updated for the last x minutes or when user is about to leave the page ?

I'm open to any other suggestion that fixes it.

**Comment by homeworkclubco** [comment]
I would love to see something like this - I've tried explaining to clients that they don't need to press save after every little change but it doesn't seem to stop them! 

I do wonder if some UX changes to the CMS admin could make some improvements here. I've made a block based site for a client which makes heavy use of objects to group things, and every time you're editing an object (lets say an image + alt text) the main visible action is "save" which will trigger a build.

My hypothesis is that a more clear "back" button or something along those lines could help in reducing excessive saves and builds

**Comment by donnersvensson** [comment]
We would love this feature, too!

Because of way to many builds on Vercel we made this workaround:

First we removed automatic builds on vercel with a vercel.json file:

`
{
  "git": {
    "deploymentEnabled": {
      "main": false
    }
  }
}
`

Then we created a "publish" collection, which triggers a Vercel Deploy Hook for the main branch, if a boolean is confirmed and the collections is saved:

`{
        name: "publish",
        label: "Publish",
        path: "content/publish",
        fields: [
          {
            type: "boolean",
            name: "publish",
            label: "Veröffentlichen",
            description: 'Please confirm and hit "Save"',
            required: false,
          },
        ],
        ui: {
          //make it a single document collection
          allowedActions: {
            create: false,
            delete: false,
          },
          // Example of beforeSubmit
          beforeSubmit: async ({
            form,
            cms,
            values,
          }: {
            form: Form;
            cms: TinaCMS;
            values: Record<string, any>;
          }) => {
            // console.log("beforeSubmit");
            await callDeployHook();
            //reload frontend
            window.location.reload();
            return {
              ...values,
              publish: false,
            };
          },
        },
      },`
      
This is no solution to all the commits in Github, but at least we do not have too many builds on Vercel and content only gets published if the editor wants it to (If you don´t have ISR enabled).

**Reply by codysaylor** [reply]
@donnersvensson nice solution, thanks for sharing. If you still want to keep the builds for code changes outside of the content folder, you can add an `ignoreCommand` value to `vercel.json` like this:

```json
{
  "ignoreCommand": "if [ \"$IS_WEBHOOK\" == \"true\" ]; then exit 0; fi; if [ \"$VERCEL_GIT_COMMIT_REF\" != \"\" ]; then bash build-non-content.sh; else exit 0; fi"
}
```

`callDeployHook()` should include the IS_WEBHOOK environment variable so it *does* trigger a build even if there are only content changes.

which would then execute a bash script called build-non-content.sh that looks like this:

```sh
# Get all changed files
CHANGED_FILES=$(git diff --name-only HEAD^ HEAD)

# Check if any files outside src/content were changed
echo "$CHANGED_FILES" | grep -v "^src/content/" > /dev/null

if [ $? -eq 0 ]; then
  echo "Changes outside src/content detected, proceeding with build"
  exit 1
else
  echo "Only changes in src/content, skipping build"
  exit 0
fi
```
